static/app/components/core/layout/styles.tsx

```diff
@@ -139,3 +143,134 @@ export function getSpacing(
+ * instead of generating CSS media queries.
+ */
+type ResponsiveValue<T> = T extends Responsive<infer U> ? U : never;
+export function useResponsivePropValue<T extends Responsive<any>>(
+  prop: T
+): ResponsiveValue<T> {
+  const activeBreakpoint = useActiveBreakpoint();
+
+  // Only resolve the active breakpoint if the prop is responsive, else ignore it.
+  if (!isResponsive(prop)) {
+    return prop as ResponsiveValue<T>;
+  }
+
+  if (Object.keys(prop).length === 0) {
+    throw new Error('Responsive prop must contain at least one breakpoint');
+  }
+
+  // If the active breakpoint exists in the prop, return it
+  if (prop[activeBreakpoint] !== undefined) {
+    return prop[activeBreakpoint];
+  }
+
+  let value: ResponsiveValue<T> | undefined;
+
+  const activeIndex = BREAKPOINT_ORDER.indexOf(activeBreakpoint);
+
+  // If we don't have an exact match, find the next smallest breakpoint
+  for (let i = activeIndex - 1; i >= 0; i--) {
+    const smallerBreakpoint = BREAKPOINT_ORDER[i]!;
+    if (prop[smallerBreakpoint] !== undefined) {
+      value = prop[smallerBreakpoint];
+      break;
+    }
+  }
+
+  // If no smaller breakpoint found, then window < smallest breakpoint, so we need to find the first larger breakpoint
+  if (value === undefined) {
+    for (let i = activeIndex + 1; i < BREAKPOINT_ORDER.length; i++) {
+      const largerBreakpoint = BREAKPOINT_ORDER[i]!;
+      if (prop[largerBreakpoint] !== undefined) {
+        value = prop[largerBreakpoint];
+        break;
+      }
+    }
+  }
+
+  return value as ResponsiveValue<T>;
+}
+
+export function useActiveBreakpoint(): Breakpoint {
+  const theme = useTheme();
+
+  const mediaQueries = useMemo(() => {
+    if (typeof window === 'undefined' || !window.matchMedia) {
+      return [];
+    }
+
+    const queries: Array<{breakpoint: Breakpoint; query: MediaQueryList}> = [];
+
+    // Iterate in reverse so that we always find the largest breakpoint
+    for (let i = BREAKPOINT_ORDER.length - 1; i >= 0; i--) {
+      const bp = BREAKPOINT_ORDER[i];
+
+      if (bp === undefined) {
+        continue;
+      }
+
+      queries.push({
+        breakpoint: bp,
+        query: window.matchMedia(`(min-width: ${theme.breakpoints[bp]})`),
+      });
+    }
+
+    return queries;
+  }, [theme.breakpoints]);
+
+  const [activeBreakpoint, setActiveBreakpoint] = useState(() => {
+    return findLargestBreakpoint(mediaQueries);
+  });
+
+  useEffect(() => {
+    if (!mediaQueries) {
+      return undefined;
+    }
+
+    function onMediaQueryChange() {
+      if (!mediaQueries) {
+        return;
+      }
+      setActiveBreakpoint(findLargestBreakpoint(mediaQueries));
+    }
+
+    for (const query of mediaQueries) {
+      query.query.addEventListener('change', onMediaQueryChange);
+    }
+
+    return () => {
+      for (const query of mediaQueries) {
+        query.query.removeEventListener('change', onMediaQueryChange);
```

Comment:
a nice trick here could be to use an `AbortSignal` instead: https://css-tricks.com/using-abortcontroller-as-an-alternative-for-removing-event-listeners/